cpp/src/lance/format/schema.cc

```diff
@@ -46,14 +46,15 @@ Field::Field(const std::shared_ptr<::arrow::Field>& field)
       extension_name_(arrow::GetExtensionName(field->type()).value_or("")),
       encoding_(pb::NONE) {
   if (is_extension_type()) {
-    auto ext_type = std::static_pointer_cast<::arrow::ExtensionType>(field->type());
+    auto ext_type = std::dynamic_pointer_cast<::arrow::ExtensionType>(field->type());
```

Comment:
the dynamic cast is for subclasses of ExtensionType right?

Reply:
`field->type()` returns `std::shared_ptr<DataType>`. `ExtensionType` is a subclass of DataType, so use dynamic cast here?